[Nomad] Simplify pul nomad to make it much easier to contribute (#6155)
pulibrary/princeton_ansible: roles/pul_nomad/README.md

* `nomad_node_pool` - pool to assign to a client VM. Right now we have `staging` and `production` - VMs in VMWare's staging infrastructure with the Nomad client software have this set to `staging` and VMs in VMWare's production infrastructure with the Nomad client software have this set to `production`.
* `nomad_host_volumes` - map of host volumes to create and make available to containers. See [Host Volumes](#host-volumes) for more. Example: 
    ```
    nomad_host_volumes:
      - name: loki
        path: '/container_data/loki'
        read_only: false
    ```
* `nomad_meta` - map of meta attributes to assign to the Nomad host. Example:
    ```
    nomad_meta:
      node_type: 'worker'
    ```
* `consul_gossip_encryption_key` - Gossip encryption key for consul/nomad server software.
* `consul_acl_management_token` Consul management token. See [Permissions & ACLs](#permissions--acls)
* `pul_nomad_management_token` Nomad management token. See [Permissions & ACLs](#permissions--acls)

## Consul

Consul is a service registration system with a DNS API.

What this means is that you can register services in it, like

> `128.4.1.2 on port 8216 is a postgres database`, and then other services can

Reviewer: You could, for example, register the ip `128.4.1.2` and port `8216` as a postgres database, and then...

There's a second config file at `/etc/dnsmasq.d/dnsmasq-10-consul` which we create from [`templates/dnsmasq/dnsmasq-10-consul.j2`](templates/dnsmasq/dnsmasq-10-consul.j2). It configures Princeton's DNS servers and the fallback to Consul for `.consul` DNS requests.

#### Debugging

To test if DNSMasq is working on a client, run `dig @127.0.0.1 -p 8600 consul.service.consul ANY` and compare it to the results of `dig consul.service.consul ANY`. If they have the same answers, DNSMasq is working.

If you need to flush the DNS Cache because IPs have changed, you can restart DNSMasq with `sudo service dnsmasq restart`

## Nomad

Nomad is a container orchestrator.

What this means is you can write job files, tell Nomad to deploy them, and Nomad
will automatically find a client VM in the cluster that can run them. If that
client goes down it will launch the container on a different client.

We install Podman on our Nomad client VMs so they can run containers without us
having to rely on Docker.

> ### Server Types

Reviewer 2: ```suggestion
### VM Types
```

### Configuration

Nomad's full configuration reference can be found in their [documentation](https://developer.hashicorp.com/nomad/docs/configuration). You may have to use the table of contents to see details of all the sections.

Every Nomad Client & Server install gets `/etc/nomad.d/base.hcl` created from [`templates/nomad/base.hcl.j2`](templates/nomad/base.hcl.j2).

Every Server install gets `/etc/nomad.d/server.hcl` created from [`templates/nomad/server.hcl.j2`](templates/nomad/server.hcl.j2).

Every Client install gets `/etc/nomad.d/client.hcl` created from [`templates/nomad/client.hcl.j2`](templates/nomad/client.hcl.j2)

The "gossip encryption key" is a key used to encrypt communication between Nomad server installations. We use the same key for both Consul & Nomad.

### Consul Interaction

Nomad uses Consul for clustering (the Nomad server installs find each other by querying
Consul) and service discovery (when you launch a job in Nomad it can register a
service in Consul that we can point to from our load balancers.)

### Host Volumes

> Host volumes are a way for containers to save data to the underlying client. It's a folder on the client which the container maps - if a job requires a certain host volume it will only ever get scheduled on a node which has that volume. Most containers have no need for persistent data on disk - they store that data in a database or something similar, but if we run a database or something like that in Nomad it might need a host volume. In those cases, we must ensure high availability through some other means (e.g "cloud" mode on Solr, clusters in Postgres, HA Mode in Redis, etc.)

Reviewer 2: ```suggestion
Host volumes are a way for containers to save data to the underlying client VM. Most containers have no need for persistent data on disk - they store that data in a database or something similar, but if we run a database or something like that in Nomad it might need a host volume. A host volume is a folder on the client which the container maps - if a job (service) requires a certain host volume it will only ever get scheduled on a node which has that volume. In those cases, because we are tying the job (service) to a specific node, we must ensure high availability through some other means (e.g "cloud" mode on Solr, clusters in Postgres, HA Mode in Redis, etc.)
```

Nomad has a useful [tutorial](https://developer.hashicorp.com/nomad/tutorials/stateful-workloads/stateful-workloads-host-volumes) about how they work.

More recent versions of Nomad have "dynamic host volumes" which would allow us to create these on the fly, not in Ansible, but we're not upgraded enough for that. Nomad has a [tutorial](https://developer.hashicorp.com/nomad/tutorials/stateful-workloads/stateful-workloads-dynamic-host-volumes) for those.

### Application Configurations

Containers deployed to the Nomad cluster may still need external systems provisioned and variables persisted to Nomad. For example an application may need a postgres database set up on our cluster, secrets installed into Nomad (which are currently in Ansible Vault), or other configuration done.

> Those things happen in [`tasks/application_configurations.yml`](tasks/application_configurations.yml).

Reviewer 2: ```suggestion
Those things happen in [`tasks/application_configurations.yml`](tasks/application_configurations.yml). Each application we deploy to Nomad has its own configuration tasks file included here.
```

This pattern may change in the future. If you want to run a playbook without running these application configurations you can do: `ansible-playbook playbooks/nomad.yml --skip-tags app_config`. You may want to do that if you're just hoping to upgrade a client or server install, change the Nomad/Consul configuration, or other tasks that are about the cluster and not the applications on it.

> Similarly, to update a single app's configuration you can do something like `ansible-playbook playbooks/nomad.yml --tags dpulc`

Reviewer 2: ```suggestion
Similarly, to update a single app's configuration you can do something like `ansible-playbook playbooks/nomad.yml --tags dpulc`, because the include tasks all have tags on them.
```

These tokens are full-access tokens that can do anything in the system. We use them in Ansible to provision more restrictive tokens, and developers can use them to navigate the system if needed.

### Consul Tokens

These are set up in `tasks/consul/acl.yml`.

Every client and server VM in the cluster has Consul installed. Some tokens are global (the same on every node), some are local (a unique token per node.) Each pul_nomad installation uses these consul tokens:

1. **DNS Token (global)**: Consul's [DNS service](https://developer.hashicorp.com/consul/docs/discover/dns) uses this token to get service information from the cluster. If this is wrong or unset, `dig @127.0.0.1 -p 8600 consul.service.consul ANY` will return no answers.
1. **Agent Token (local)**: Consul's agent (the process run via `sudo service consul start`, uses this to communicate with the rest of the cluster and synchronize services. This uses a "[node identity](https://developer.hashicorp.com/consul/docs/secure/acl/role#node-identities)" token, which has a default policy.

> 1. **Nomad Server Token (global)**: Allows Nomad's server installs to mesh, manage ACLs, and update Consul services. Every Nomad Server VM has this configured. Defined in [nomad/configure_server.yml](tasks/nomad/configure_server.yml)
> 1. **Nomad Client Token (global)**: Allows Nomad's client installs to update Consul services as they're deployed into Nomad. Every Nomad Client VM install has this configured. Defined in [nomad/configure_client.yml](tasks/nomad/configure_client.yml)

Reviewer 2: ```suggestion
1. **Nomad Server Token (global)**: Nomad's server uses this token to talk to Consul to learn about each other's existence, manage ACLs, and update Consul services. Every Nomad Server VM has this configured. Defined in [nomad/configure_server.yml](tasks/nomad/configure_server.yml)
1. **Nomad Client Token (global)**: Nomad's client uses this token to update Consul services as they're deployed into Nomad. Every Nomad Client VM install has this configured. Defined in [nomad/configure_client.yml](tasks/nomad/configure_client.yml)
```

### Nomad Tokens

We don't currently have any special nomad tokens, just the management token. We create that token when we [bootstrap](https://developer.hashicorp.com/nomad/tutorials/access-control/access-control-bootstrap) Nomad's ACL system in [nomad/server.yml](tasks/nomad/server.yml)

> # Generating Variables
> 
> Some variables need to be generated and stored in the vault. To generate new
> ones for consul (`consul_gossip_encryption_key`) you can run:

Reviewer 2: ```suggestion
# Refreshing the Gossip Encryption Key

To refresh the gossip encryption key for consul (`consul_gossip_encryption_key`), generate a new key and save it to the vault. To generate the key, run:
```

`nomad-host-*.princeton.edu`

Clients run jobs. In our case, containers executed by Podman. These are named
`nomad-client-*.princeton.edu`

### Configuration

Nomad's full configuration reference can be found in their [documentation](https://developer.hashicorp.com/nomad/docs/configuration). You may have to use the table of contents to see details of all the sections.

Every Nomad Client & Server gets `/etc/nomad.d/base.hcl` created from [`templates/nomad/base.hcl.j2`](templates/nomad/base.hcl.j2).

Every Server gets `/etc/nomad.d/server.hcl` created from [`templates/nomad/server.hcl.j2`](templates/nomad/server.hcl.j2).

Every Client gets `/etc/nomad.d/client.hcl` created from [`templates/nomad/client.hcl.j2`](templates/nomad/client.hcl.j2)

The "gossip encryption key" is a key used to encrypt communication between Nomad servers. We use the same key for both Consul & Nomad.

### Consul Interaction

Nomad uses Consul for clustering (the Nomad servers find each other by querying
Consul) and service discovery (when you launch a job in Nomad it can register a
service in Consul that we can point to from our load balancers.)

### Host Volumes

> Host volumes are a way for containers to save data to the underlying client. It's a folder on the client which the container maps - if a job requires a certain host volume it will only ever get scheduled on a node which has that volume.

Reviewer 2: ```suggestion
Host volumes are a way for containers to save data to the underlying client VM. It's a folder on the client VM which the container maps - if a job requires a certain host volume it will only ever get scheduled on a node which has that volume.
```
